File: arrays/three_sum/go/solution_two_pointers.go
package main

import "sort"

func threeSumTwoPointers(nums []int) [][]int {
	result := make([][]int, 0)

	// Предварительно сортируем слайс по возрастанию
	sort.Ints(nums)

	for i, num := range nums {
		// Не проверяем текущее число, если оно такое же, как и предыдущее, потому для него мы получим такой же результат.
		if i != 0 && nums[i-1] == num {
			continue
		}

		left := i + 1
		right := len(nums) - 1

		for left < right {
			sum := num + nums[left] + nums[right]

			if sum == 0 {
				result = append(result, []int{num, nums[left], nums[right]})
				right -= 1
				left += 1

				for left < right && nums[left] == nums[left-1] {
					left += 1
				}
			} else if sum > 0 {
				right -= 1
			} else {
				left = left + 1
			}
		}
	}

	return result
}

File: arrays/three_sum/ts/solution_two_pointers.ts
export const threeSumTwoPointers = (nums: number[]): number[][] => {
    const result: number[][] = []

    // Предварительно сортируем слайс по возрастанию
    nums.sort()

    for (let i = 0; i < nums.length; i++) {
        const num = nums[i]

        // Не проверяем текущее число, если оно такое же, как и предыдущее, потому для него мы получим такой же результат.
        if (i != 0 && nums[i-1] == num) {
            continue
        }

        let left = i + 1
        let right = nums.length - 1

        while (left < right) {
            const sum = num + nums[left] + nums[right]

            if (sum == 0) {
                result.push([num, nums[left], nums[right]])
                right -= 1
                left += 1

                while (left < right && nums[left] == nums[left-1]) {
                    left += 1
                }
            } else if (sum > 0) {
                right -= 1
            } else {
                left = left + 1
            }
        }
    }

    return result
}
